### ISSUE TITLE
Incorrect INT8 quantized BERT model prediction on AVX2 only CPU

### ISSUE THREAD

**Issue by jaroslavgratz** [microsoft/onnxruntime#6004]
Inference of a BERT INT8 quantized model gives incorrect result on AVX2 only (no AVX-512 support) CPU. Please note this issue is related to INT8 quantization and AVX2 CPU only . INT8 quantized model works correctly on AVX-512 CPU. UINT8 (unsigned int) quantized model works correctly on both AVX2 CPU and AVX-512 CPU.

It is possible to reproduce this bug on any CPU by running inference under valgrind (valgrind does't support AVX-512). Moreover valgrind reports access to uninitialized variables during inference (likely to be related to incorrect prediction result):

```
Conditional jump or move depends on uninitialised value(s)
   at 0x515C8E7: onnxruntime::DynamicQuantizeLinear<unsigned char>::Compute(onnxruntime::OpKernelContext*) const (in /home/jag/bootstrap/local/lib/libonnxruntime.so.1.5.3)
   by 0x5346538: onnxruntime::SequentialExecutor::Execute(onnxruntime::SessionState const&, std::vector<int, std::allocator<int> > const&, std::vector<OrtValue, std::allocator<OrtValue> > const&, std::vector<int, std::allocator<int> > const&, std::vector<OrtValue, std::allocator<OrtValue> >&, std::unordered_map<unsigned long, std::function<onnxruntime::common::Status (onnxruntime::TensorShape const&, OrtMemoryInfo const&, OrtValue&, bool&)>, std::hash<unsigned long>, std::equal_to<unsigned long>, std::allocator<std::pair<unsigned long const, std::function<onnxruntime::common::Status (onnxruntime::TensorShape const&, OrtMemoryInfo const&, OrtValue&, bool&)> > > > const&, onnxruntime::logging::Logger const&) (in /home/jag/bootstrap/local/lib/libonnxruntime.so.1.5.3)
   by 0x533233B: onnxruntime::utils::ExecuteGraphImpl(onnxruntime::SessionState const&, onnxruntime::FeedsFetchesManager const&, std::vector<OrtValue, std::allocator<OrtValue> > const&, std::vector<OrtValue, std::allocator<OrtValue> >&, std::unordered_map<unsigned long, std::function<onnxruntime::common::Status (onnxruntime::TensorShape const&, OrtMemoryInfo const&, OrtValue&, bool&)>, std::hash<unsigned long>, std::equal_to<unsigned long>, std::allocator<std::pair<unsigned long const, std::function<onnxruntime::common::Status (onnxruntime::TensorShape const&, OrtMemoryInfo const&, OrtValue&, bool&)> > > > const&, ExecutionMode, bool const&, onnxruntime::logging::Logger const&, bool) (in /home/jag/bootstrap/local/lib/libonnxruntime.so.1.5.3)
   by 0x5334318: onnxruntime::utils::ExecuteGraph(onnxruntime::SessionState const&, onnxruntime::FeedsFetchesManager&, std::vector<OrtValue, std::allocator<OrtValue> > const&, std::vector<OrtValue, std::allocator<OrtValue> >&, ExecutionMode, bool const&, onnxruntime::logging::Logger const&, bool) (in /home/jag/bootstrap/local/lib/libonnxruntime.so.1.5.3)
   by 0x4EAF207: onnxruntime::InferenceSession::Run(OrtRunOptions const&, std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > const&, std::vector<OrtValue, std::allocator<OrtValue> > const&, std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > const&, std::vector<OrtValue, std::allocator<OrtValue> >*, std::vector<OrtDevice, std::allocator<OrtDevice> > const*) (in /home/jag/bootstrap/local/lib/libonnxruntime.so.1.5.3)
   by 0x4E7CF63: OrtApis::Run(OrtSession*, OrtRunOptions const*, char const* const*, OrtValue const* const*, unsigned long, char const* const*, unsigned long, OrtValue**) (in /home/jag/bootstrap/local/lib/libonnxruntime.so.1.5.3)
   by 0x13A2A6: Run (onnxruntime_cxx_inline.h:475)
   by 0x13A2A6: Run (onnxruntime_cxx_inline.h:466)
```

**System information**
- OS Platform and Distribution: Debian Buster
- ONNX Runtime installed from (source or binary): source
- ONNX Runtime version: 1.5.3
- Python version: 3.7.3
- GCC/Compiler version (if compiling from source): GCC 8.3.0

**Comment by snnn** [comment]
Even with AVX-512, when I ran the unit tests, I saw a lot of similar warnings. 

```
==3855== Conditional jump or move depends on uninitialised value(s)
==3855==    at 0x7FB6E7: CmpHelperEQ<unsigned char, unsigned char> (gtest.h:1527)
==3855==    by 0x7FB6E7: Compare<unsigned char, unsigned char, 0u> (gtest.h:1554)
==3855==    by 0x7FB6E7: void onnxruntime::test::Check<unsigned char>(onnxruntime::test::OpTester::Data const&, onnxruntime::Tensor const&, std::string const&) (provider_test_utils.cc:66)
==3855==    by 0x80166D: CheckDispatch<float, double, unsigned char, short unsigned int, unsigned int, long unsigned int, signed char, short int, int, long int, std::basic_string<char, std::char_traits<char>, std::allocator<char> >, onnxruntime::MLFloat16, onnxruntime::BFloat16> (provider_test_utils.cc:272)
==3855==    by 0x80166D: CheckDispatch<bool, float, double, unsigned char, short unsigned int, unsigned int, long unsigned int, signed char, short int, int, long int, std::basic_string<char, std::char_traits<char>, std::allocator<char> >, onnxruntime::MLFloat16, onnxruntime::BFloat16> (provider_test_utils.cc:272)
==3855==    by 0x80166D: onnxruntime::test::Check(onnxruntime::test::OpTester::Data const&, onnxruntime::Tensor const&, std::string const&) (provider_test_utils.cc:289)
==3855==    by 0x80453B: std::vector<OrtValue, std::allocator<OrtValue> > onnxruntime::test::OpTester::ExecuteModel<onnxruntime::InferenceSession>(onnxruntime::Model&, onnxruntime::InferenceSession&, onnxruntime::test::OpTester::ExpectResult, std::string const&, OrtRunOptions const*, std::unordered_map<std::string, OrtValue, std::hash<std::string>, std::equal_to<std::string>, std::allocator<std::pair<std::string const, OrtValue> > > const&, std::vector<std::string, std::allocator<std::string> > const&, std::string const&) (provider_test_utils.cc:625)
==3855==    by 0x8072AF: onnxruntime::test::OpTester::Run(onnxruntime::SessionOptions, onnxruntime::test::OpTester::ExpectResult, std::string const&, std::unordered_set<std::string, std::hash<std::string>, std::equal_to<std::string>, std::allocator<std::string> > const&, OrtRunOptions const*, std::vector<std::unique_ptr<onnxruntime::IExecutionProvider, std::default_delete<onnxruntime::IExecutionProvider> >, std::allocator<std::unique_ptr<onnxruntime::IExecutionProvider, std::default_delete<onnxruntime::IExecutionProvider> > > >*, onnxruntime::Graph::ResolveOptions const&) (provider_test_utils.cc:864)
==3855==    by 0x8087D3: onnxruntime::test::OpTester::Run(onnxruntime::test::OpTester::ExpectResult, std::string const&, std::unordered_set<std::string, std::hash<std::string>, std::equal_to<std::string>, std::allocator<std::string> > const&, OrtRunOptions const*, std::vector<std::unique_ptr<onnxruntime::IExecutionProvider, std::default_delete<onnxruntime::IExecutionProvider> >, std::allocator<std::unique_ptr<onnxruntime::IExecutionProvider, std::default_delete<onnxruntime::IExecutionProvider> > > >*, ExecutionMode, onnxruntime::Graph::ResolveOptions const&) (provider_test_utils.cc:657)
==3855==    by 0x8C4505: onnxruntime::test::QuantizeLinearMatmulOpTest_QLinearMatMul3D_U8U8_Test::TestBody() (quantize_linear_matmul_test.cc:45)
==3855==    by 0x11B14E2: HandleSehExceptionsInMethodIfSupported<testing::Test, void> (gtest.cc:2433)
==3855==    by 0x11B14E2: void testing::internal::HandleExceptionsInMethodIfSupported<testing::Test, void>(testing::Test*, void (testing::Test::*)(), char const*) (gtest.cc:2469)
==3855==    by 0x11A0D5F: testing::Test::Run() (gtest.cc:2509)
==3855==    by 0x11A0E80: testing::TestInfo::Run() (gtest.cc:2684)
==3855==    by 0x11A0F44: testing::TestSuite::Run() (gtest.cc:2816)
==3855==    by 0x11A9E76: testing::internal::UnitTestImpl::RunAllTests() (gtest.cc:5338)
==3855==    by 0x11AA01B: HandleSehExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool> (gtest.cc:2433)
==3855==    by 0x11AA01B: HandleExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool> (gtest.cc:2469)
==3855==    by 0x11AA01B: testing::UnitTest::Run() (gtest.cc:4928)
==3855==    by 0x4E0165: RUN_ALL_TESTS (gtest.h:2473)
==3855==    by 0x4E0165: main (test_main.cc:67)
```

**Comment by snnn** [comment]
Hi @jaroslavgratz , would you please give us a unit test or model with test data to repro the issue?

**Comment by snnn** [comment]
I tried to run all our unit test on a machine with AVX2(but without AVX512), then I searched "DynamicQuantizeLinear", I didn't find the similar thing.  So maybe we don't have a unit test to cover it, or it only shows up in some special software environments.

**Comment by yufenglee** [comment]
@snnn, we have unit test for DynamicQuantizeLinear: https://github.com/microsoft/onnxruntime/blob/9ec1ed42a809170b87474f5822c4557101812399/onnxruntime/test/providers/cpu/tensor/dynamic_quantize_linear_test.cc .

**Comment by jaroslavgratz** [comment]
See please int8 quantized model in attachement. Code to reproduce:
```
#include <onnxruntime/core/session/onnxruntime_cxx_api.h>
#include <iostream>

int main(int argc, char** argv)
{
    Ort::Env env(ORT_LOGGING_LEVEL_WARNING, "onnxruntime");

    Ort::SessionOptions options;
    options.SetIntraOpNumThreads(1);
    options.SetInterOpNumThreads(1);
    Ort::Session session(env, "model.onnx", options);

    std::vector<const char*> inputNames = {"input_ids", "attention_mask", "token_type_ids" };
    std::vector<const char*> outputNames = {"output"};

    // dummy input
    std::vector<int64_t> ids(128, 1);
    std::vector<int64_t> attention_mask(128, 1);
    std::vector<int64_t> type_ids(128, 1);

    int64_t ids_dims[2] = {1, static_cast<int64_t>(ids.size())};
    int64_t type_ids_dims[2] = {1, static_cast<int64_t>(type_ids.size())};
    int64_t attention_mask_dims[2] = {1, static_cast<int64_t>(attention_mask.size())};

    auto memory_info = Ort::MemoryInfo::CreateCpu(OrtArenaAllocator, OrtMemTypeDefault);

    std::vector<Ort::Value> inputs;
    inputs.emplace_back(Ort::Value::CreateTensor<int64_t>(memory_info, ids.data(), ids.size(), ids_dims, 2));
    inputs.emplace_back(Ort::Value::CreateTensor<int64_t>(memory_info, attention_mask.data(), attention_mask.size(), attention_mask_dims, 2));
    inputs.emplace_back(Ort::Value::CreateTensor<int64_t>(memory_info, type_ids.data(), type_ids.size(), type_ids_dims, 2));

    auto result = session.Run(Ort::RunOptions{nullptr}, inputNames.data(), inputs.data(), inputNames.size(), outputNames.data(), outputNames.size());

    std::cout << "RESULT: " << *result.front().GetTensorData<float>() << std::endl;

    return 0;
}
```
Result on Xeon(R) Silver 4210 CPU (AVX-512 including VNNI): 1.34879
Result on AMD EPYC 7401P (no AVX-512): 1.43256
Result under valgrind (no AVX-512): 1.43256

Please note that I don't see valgrind errors about access to uninitialized data in this simplified test, however inference results are different on AVX2 and AVX-512 CPUs.
[model.zip](https://github.com/microsoft/onnxruntime/files/5634050/model.zip)

**Comment by yufenglee** [comment]
@jaroslavgratz , your issue is caused by saturation that is explained in another thread: https://github.com/microsoft/onnxruntime/issues/5849#issuecomment-733178450. Could you please try the reduce_range option when quantizing your model to see if the issue is resolved?

**Comment by jaroslavgratz** [comment]
Thank you for your explanation. In fact, different prediction on non-VNNI hardware is not an issue for me. I can ensure a model will run on VNNI machines in production. However, model accuracy loss can be quite high on non-VNNI hardware and that can be confusing for users not familiar with AVX instuction sets differences. Please consider use of a 32-bit accumulator for 8-bit multiply results for non-VNNI algorithm, that can improve user experience significantly.
